### ISSUE TITLE
RFC: simple support for other testing tools

### ISSUE THREAD

**Issue by BackEndTea** [infection/infection#96]
I'd like to propose a simple way to support other testing tools.

Allow additional commands to be ran, if a mutant is not killed/timed out. If these commands terminate with a non 0 exit code, it means the mutant was killed.

For example given the follwing ```infection.json```

```json
{
    "timeout": 20,
    "source": {
        "directories": [
            "classes"
        ]
    },
    "logs": {
        "text": "infection-log.txt"
    },
    "phpUnit": {
        "configDir": "tests"
    }
}
```
It could be changed to something like this:
```json
{
    "timeout": 20,
    "source": {
        "directories": [
            "classes"
        ]
    },
    "logs": {
        "text": "infection-log.txt"
    },
    "phpUnit": {
        "configDir": "tests"
    },
    "fallback" : [
        "vendor/bin/other-php-tester",
        "./tester.sh"
    ]
}
```

This would slow down the tests of course and users are responsible for adding ```--stop-on-failure``` like flags, but it would give the users more security in the results of infection.

**Comment by theofidry** [comment]
That's an interesting idea and although not ideal performance-wise it would provide a basic support for any kind of testing tool

**Comment by BackEndTea** [comment]
@theofidry 
I agree that its really not ideal preformance wise. But depending on your use case that should not really matter.

I don't think you would run it like this on CI for example. But you may run it like this every once in a while locally to see where you can improve your test suite.

**Comment by sidz** [comment]
I don't any sense of doing fallbacks which will run other tools. For example it easily can be configured by CI or via composer.json like we have in `infection`.

Like `composer analyze` will run `cs.check` and `phpstan`.

Maybe I missed smth useful.

**Comment by borNfreee** [comment]
I quite like the idea as it adds a really simple way of running several test frameworks to kill the mutant.

Could you please describe your use case? What particular tools do you use except PHPUnit that you want to run as fallbacks?

Also, it would be great to know how many people are interested in this feature. May be we need to create a poll in twitter ;)

Why I have these concerns is because it's quite tricky to implement this feature, we need to rewrote logic of ParallelProcessRunner.php, Mutant (currently it has only 1 process to run) and our listeners (to collect metrics, to advance progress bar, etc.).

**Comment by theofidry** [comment]
Besides how we implement it behind, I'm also not convinced the suggested config is the best way to go for it either. I don't have a better proposal for it yet though :)

**Comment by BackEndTea** [comment]
I would use it to just run php unit again, since `@covers` annotations can mess with the results sometimes, especially for integration tests. And also for tests like behat.

>Besides how we implement it behind, I'm also not convinced the suggested config is the best way to go for it either. I don't have a better proposal for it yet though :)

The config as i proposed it is probably a bad way to implement it. Cli flags might work better

**Comment by Ocramius** [comment]
> Could you please describe your use case? What particular tools do you use except PHPUnit that you want to run as fallbacks?

Mentioned this [today](https://twitter.com/Ocramius/status/1108446593655754753), and what I'd love to run is `vendor/bin/psalm <path-to-the-mutated-source-file>` or `vendor/bin/phpstan analyse <path-to-the-mutated-source-file>`.

Whilst inefficient, this automatically squishes mutants that either:

 * deviate from a coding standard (most "good" coding standards enforce stricter/simpler language)
 * make the code unreasonable from a type perspective, such as function call removals that change the type of an expression

This would also allow becoming much more aggressive with mutations: mutants that wouldn't make much sense (edge cases, or things that we consider "excessive") could become viable again, since this sort of tooling would rule out most cases where obvious false positives arise.

**Comment by theofidry** [comment]
>make the code unreasonable from a type perspective, such as function call removals that change the type of an expression

Don't you think that trying to make the mutant smarter via a static tool analyser (e.g. relying on PHPStan AST parsing which provides more information than the row AST) would be better to achieve that?

**Comment by Ocramius** [comment]
@theofidry possibly, but that's another massive amount of steps down the line. In a language like PHP, changing expression type is perfectly valid in many contexts, so "dumb mutations" are actually "as dumb as the developer"

**Comment by maks-rafalko** [comment]
I would like to discuss this one again.

1. @theofidry what are your concerns about the proposed configuration?

2. as Theo already mentioned, having smarter Mutators is indeed a better approach, but I don't think these 2 approaches are exclusive. 

3. More and more we are asked to implement some features that are impossible at the moment as we don't know variable types. It's just a matter of time and I'm pretty sure we will have to implement it at some point. Is anybody aware of PHPStan's implementation? Is it reusable?

---

I would vote for adding this particular feature with running `fallback` script to `0.14.0`

**Comment by BackEndTea** [comment]
Using psalm and phpstan with the code generated by infection will generate errors (killed mutatants) when it shouldn't. At least with most configurations.

Most of these will be due to variables not being used anymore, or if statements being always true.

I know stryker ran into issues that are somewhat alike for mutating typescript code, so they set specific compiler options, (to ignore these issues). We could also implement this in infection, but it would couple us to psalm/phpstan.

I think suppport for these tools/ configuring them should be part of an (optional) plugin, maintained by the infection team or not. That would mean we need to actually set up this plugin system.

If we wish to get this in `0.14.0` we should do so with a big disclaimer that the users is responsible for making sure the fallback option doesn't error out on errors that it shouldn't.

**Comment by maks-rafalko** [comment]
> Most of these will be due to variables not being used anymore, or if statements being always true.

Hm.. makes sense.

@Ocramius could you please give a couple of different examples where running psalm/phpstan after PHPUnit during mutation process would be useful? To get a better understanding of your cases.

**Comment by Ocramius** [comment]
Unused variables and code style aren't really 'incorrect" code, but passing apples where oranges are expected is certainly a problem, and mutants must be killed in that case.

A typical example of a mutant that I don't really wish to test is `array_values()` removal, because my code declares `array<int, string>` returned and `array<string>` is operated upon, mutations lead to invalid code that still passes tests, but doesn't make sense according to type declarations.

In practice, mutations that break type checks are exactly the same as mutants that fail/crash unit tests.

As for implementation, I think that using a plugin/hook system for experimental usage would be good for a start, while coupling to a tool may be dangerous if done right away.

**Comment by sanmai** [comment]
I really wish we could move forward with #669. That should truly open Infection greatest inventions for use in other tools, so that Infection would not have to bother about other testing tools that much. They could be dealt with in separate applications, or even have mutation testing integrated into them.
